Debug xml dsig

fisx opened the pull request:
These changes make XML signature validation more robust, and also more correct.

Tasks:

- [x] capture stdout, stderr noise from libxml2 and feed it into an error that is returned;
- [x] make sure that the signed subtrees are self-contained by applying an extra round of canonicalization;
- [x] fix test suite (currently: `"failed to canonicalize input: xmlReadMemory: illegal operation (Inappropriate ioctl for device)"`)

fisx committed:
Cleanup

fisx committed:
Fix: remove impossible error case.

fisx committed:
Cleanup

fisx committed:
Cleanup

fisx committed:
Fix: do not lose namespace declarations when cutting out signed subtree.

fisx committed:
Fix: capture stdout, stderr and translate into locatable error.

fisx committed:
Cleanup

fisx committed:
Fixup 64b0562476f9c1277bbf246114c908665b0fc750

fisx committed:
Haddocks

ChrisPenner commented on SAML2/XML/Signature.hs:
```diff
@@ -242,56 +246,109 @@ _verifySignatureOld pks xid doc = do
+  signedSubtree :: HXT.XmlTree
+    <- do
+        mdoc' <- liftIO . try $ fixNamespaces doc
+        doc' <- case mdoc' of
```
This case can be simplified to `fromRight`

ChrisPenner commented on SAML2/XML/Signature.hs:
```diff
@@ -242,56 +246,109 @@ _verifySignatureOld pks xid doc = do
+        doc' <- case mdoc' of
+          Right x
+            -> pure x
+          Left (err :: SomeException)
```
It'd be nice to move the `SomeException` type annotation up into `try @SomeException` so the exception type stays close to the `try` and makes it clear that we plan to catch ALL exceptions 😄

ChrisPenner commented on SAML2/XML/Signature.hs:
```diff
@@ -242,56 +246,109 @@ _verifySignatureOld pks xid doc = do
+          Left err -> throwError . SignatureParseError $ show err
+          Right v -> pure v
+
+  signedInfoElem :: BS.ByteString
```
Some docstring or comments explaining the magic xPaths would be nice 😄

fisx replied:
Agreed, but i would have to figure out what that code does first...  (It's not mine)

ChrisPenner commented on SAML2/XML/Signature.hs:
```diff
@@ -242,56 +246,109 @@ _verifySignatureOld pks xid doc = do
+      $ mapM (`verifyReference` signedSubtree) (signedInfoReference signedInfoTyped)
+
+  unless (all isRight referenceChecks) $
+    throwError . SignatureVerifyBadReferences $ show (signedInfoReference signedInfoTyped, referenceChecks)
```
If there's a left in the referenceChecks should we possibly `sequence` it to collect the error out? Or is it more descriptive to print everything out like this?

fisx replied:
It's a list of `Either`s, and I want to keep track of how many errors there are, and which precisely.

I agree it looks weird, but I can't think of a better way to do it.  Wanna try something and add a commit?

ChrisPenner replied:
In that case; how about `show (signedInfoReference signedInfoTyped, lefts referenceChecks)` instead? `lefts` will just grab all the errors from the list of Eithers

ChrisPenner commented on SAML2/XML/Signature.hs:
```diff
@@ -242,56 +246,109 @@ _verifySignatureOld pks xid doc = do
+capture' :: String -> IO a -> IO a
+capture' actionName action = hCapture [stdout, stderr] action >>= \case
+  ("", !out) -> pure out
+  (noise, _) -> throwIO . ErrorCall $ actionName <> ": " <> noise
```
If any noise on stdout indicates an error; it's probably helpful to have the output of stderr appended here too.

fisx replied:
i think that's what this does, no?

(thanks for the other comments, more changes coming up!)

ChrisPenner replied:
Ahh; I see! I misunderstood; It looked at a glance that the tuple matched up with `[stdout, stderr]`; but in hindsight that does't make sense 👍

ChrisPenner commented:
It's all greek to me 🤷‍♂️ 

Did this not change any behaviour in any of the tests? If that's the case we should almost certainly add some new tests to capture whatever was wrong before this change.

fisx committed:
Review

fisx commented:
> It's all greek to me 
> 
> Did this not change any behaviour in any of the tests? If that's the case we should almost certainly add some new tests to capture whatever was wrong before this change.

It does make one test fail.  See PR description.  :-)

To reproduce, run `stack test --fast` and follow the error.

fisx commented:
oh, and:

> It's all greek to me

i've stared at this for over a day now, and not for the first time, and i feel the same way.  :-).  i don't think there is a way to implement xml:dsig though that would make the reader of the code feel any better about the world.  that's just how xml standards are.

fisx committed:
Review

fisx commented:
> `xmlReadMemory: illegal operation (Inappropriate ioctl for device)`

i think i know what's going on here.

Since libxml2 doesn't return errors, but spills them on stdout, stderr, I use [hCapture](http://hackage.haskell.org/package/silently-1.2.5/docs/System-IO-Silently.html) that reads those into a string and returns it with together with the value.

`hCapture` seems to change the type of the file handle (from what to what?).  libxml2 not only spills stuff on stderr, but also performs a ioctl operation on stderr (or stdout) that is supported before `hCapture`, but not after.  hence, [ENOTTY](https://www.gnu.org/software/libc/manual/html_node/Error-Codes.html).

The first solution I can think of is to trap and ignore this error `c14n`.

fisx committed:
Cleanup (polish error messages some more)

fisx committed:
Cleanup

fisx committed:
Fix: a better approach to getting a signed subtree.

fisx committed:
Fix: how xpath appearently works.

fisx committed:
Make libxml2 parser wrapper handle stderr noise and errors better.

fisx committed:
Cleanup

fisx committed:
Fix: test suite.

fisx committed:
Fix: remove broken test case.

I think this test had some issue with whitespace being changed
after signing, and canonicalization does not touch whitespace
that much.  I am removing it because it is very likely that this
is the issue, and very hard to robustly confirm it.

fisx committed:
Add test case: idempotency of canonicalization.

fisx commented:
ready for review!

neongreen commented on SAML2/XML.hs:
```diff
@@ -100,13 +101,24 @@ samlToDoc = head
+--
+-- TODO: change this function to return 'HXT.XmlTrees' and replace 'samlToDoc' with it.
+samlToDoc' :: XP.XmlPickler a => a -> HXT.XmlTree
+samlToDoc' = head . getChildren . head
```
`samlToDoc' = head . getChildren . samlToDoc` ?

neongreen replied:
Or do you want it to remain like it is so that it would be easier to get rid of `samlToDoc` in the future?

fisx replied:
yes and yes.

neongreen commented on test/XML/Signature.hs:
```diff
@@ -302,11 +306,6 @@ examples = zipWith ($) xs [1..]
   where xs :: [Int -> VerifyExample]
         xs =
           [ VerifyExample
-            "<ds:KeyInfo xmlns:ds=\"http://www.w3.org/2000/09/xmldsig#\"><ds:X509Data><ds:X509Certificate>MIIDpDCCAoygAwIBAgIGAWOMMryDMA0GCSqGSIb3DQEBCwUAMIGSMQswCQYDVQQGEwJVUzETMBEGA1UECAwKQ2FsaWZvcm5pYTEWMBQGA1UEBwwNU2FuIEZyYW5jaXNjbzENMAsGA1UECgwET2t0YTEUMBIGA1UECwwLU1NPUHJvdmlkZXIxEzARBgNVBAMMCmRldi02MDc2NDgxHDAaBgkqhkiG9w0BCQEWDWluZm9Ab2t0YS5jb20wHhcNMTgwNTIzMDg1MTA1WhcNMjgwNTIzMDg1MjA1WjCBk ...
```
Why was this removed?

fisx replied:
see commit message.

neongreen commented on SAML2/XML/LibXML2.hsc:
```diff
@@ -63,16 +65,55 @@ newtype Doc = Doc{ unDoc :: ForeignPtr XMLDoc }
-      xmlReadMemory p (fromIntegral l) nullPtr nullPtr (#{const XML_PARSE_NOENT} .|. #{const XML_PARSE_DTDLOAD} .|. #{const XML_PARSE_DTDATTR} .|. #{const XML_PARSE_NONET} .|. #{const XML_PARSE_COMPACT})
-  newDoc d
+  resetErrno
+  let config :: CInt
```
Would be nice to have comments for those flags, because it's not obvious from the names what they do.

fisx replied:
it's not my code, and i only touched these to make it slightly easier to read.  also, i would argue this information belongs into the libxml2 manual.

neongreen commented on SAML2/XML/Signature.hs:
```diff
@@ -242,56 +247,115 @@ _verifySignatureOld pks xid doc = do
+--
+-- how does this work?:
+--   * dig for the subtree of the input with an ID attribute containing xid (the "signed subtree")
+--   * parse the 'Sigature' subtree in that subtree (we only do envelopped signatures)
```
```suggestion
--   * parse the 'Signature' subtree in that subtree (we only do envelopped signatures)
```

neongreen commented:
I can't really vouch for the logic here. Having more tests for happy and unhappy paths would be nice: good signature, good signature but non-canonical encoding, bad signature, missing signature.

Commit:
Cleanup

Co-Authored-By: fisx <mf@zerobuzz.net>

fisx commented:
> I can't really vouch for the logic here. Having more tests for happy and unhappy paths would be nice: good signature, good signature but non-canonical encoding, bad signature, missing signature.

i agree, but i won't do it in this PR.  note that some of the tests you suggest are in saml2-web-sso (or even in wire-server, i would have to read up on the details).

i will merge this now, if any of the answers i give are not enough please yell and i'll handle that in separate PRs.

neongreen commented:
Okay

On Wed, Feb 13, 2019, 21:04 fisx <notifications@github.com wrote:

> Merged #4 <https://github.com/wireapp/hsaml2/pull/4> into master.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/wireapp/hsaml2/pull/4#event-2137794794>, or mute the
> thread
> <https://github.com/notifications/unsubscribe-auth/ABc-arhE4SQT0p5wBW5i5ToCGnby0oWHks5vNG_FgaJpZM4a18E7>
> .
>
